File: Algorithms/0016_3Sum_Closest/Java/3SumClosestSolution1.java

// Space: O(1)
// Time: O(n^2)

class Solution {
    public int threeSumClosest(int[] nums, int target) {
        if (nums.length == 3) return Arrays.stream(nums).sum();
        Arrays.sort(nums);
        int res = nums[0] + nums[1] + nums[2];
        for (int i = 1; i < nums.length - 1; i++) {
            int left = 0, right = nums.length - 1;
            while (left != i && right != i) {
                if (Math.abs(nums[left] + nums[i] + nums[right] - target) < Math.abs(res - target)) {
                    res = nums[left] + nums[i] + nums[right];
                }
                if (nums[left] + nums[i] + nums[right] == target) return res;
                if (nums[left] + nums[i] + nums[right] < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }
        return res;
    }
}

File: Algorithms/0016_3Sum_Closest/Python/3Sum_Closest_Solution_1.py

# Space: O(1)
# Time: O(n^2)

class Solution:
    def threeSumClosest(self, nums, target):
        length = len(nums)
        if length == 3: return sum(nums)
        nums = sorted(nums)
        res = sum(nums[:3])
        for i in range(length - 2):
            left, right = i + 1, length - 1
            while left < right:
                if abs((nums[i] + nums[left] + nums[right]) - target) < abs(res - target):
                    res = nums[i] + nums[left] + nums[right]
                if nums[i] + nums[left] + nums[right] == target: return res
                if nums[i] + nums[left] + nums[right] < target: left += 1
                if nums[i] + nums[left] + nums[right] > target: right -= 1

        return res




